peer/abstractlist/list.go

```diff
@@ -0,0 +1,592 @@
+// one becomes available.
+//
+// The list is a suitable basis for concrete implementations like round-robin.
+type List struct {
+	lock sync.RWMutex
+	once *lifecycle.Once
+
+	name string
+
+	peers              map[string]*peerFacade
+	offlinePeers       map[string]peer.Identifier
+	implementation     Implementation
+	peerAvailableEvent chan struct{}
+	transport          peer.Transport
+
+	noShuffle bool
+	failFast  bool
+	randSrc   rand.Source
+}
+
+// Name returns the name of the list.
+func (pl *List) Name() string { return pl.name }
+
+// Transport returns the underlying transport for retaining and releasing peers.
+func (pl *List) Transport() peer.Transport { return pl.transport }
+
+// Update applies the additions and removals of peer Identifiers to the list
+// it returns a multi-error result of every failure that happened without
+// circuit breaking due to failures.
+func (pl *List) Update(updates peer.ListUpdates) error {
+	if len(updates.Additions) == 0 && len(updates.Removals) == 0 {
+		return nil
+	}
+
+	pl.lock.Lock()
+	defer pl.lock.Unlock()
+
+	if !pl.once.IsRunning() {
+		return pl.updateOffline(updates)
+	}
+	return pl.updateOnline(updates)
+}
+
+func (pl *List) updateOnline(updates peer.ListUpdates) error {
+	var errs error
+	for _, id := range updates.Removals {
+		errs = multierr.Append(errs, pl.remove(id))
+	}
+
+	add := updates.Additions
+	if !pl.noShuffle {
+		add = shuffle(pl.randSrc, add)
+	}
+
+	for _, id := range add {
+		errs = multierr.Append(errs, pl.add(id))
+	}
+	return errs
+}
+
+func (pl *List) updateOffline(updates peer.ListUpdates) error {
+	var errs error
+	for _, id := range updates.Removals {
+		errs = multierr.Append(errs, pl.removeOffline(id))
+	}
+	for _, id := range updates.Additions {
+		errs = multierr.Append(errs, pl.addOffline(id))
+	}
+	return errs
+}
+
+// Start notifies the List that requests will start coming
+func (pl *List) Start() error {
+	return pl.once.Start(pl.start)
+}
+
+func (pl *List) start() error {
+	pl.lock.Lock()
+	defer pl.lock.Unlock()
+
+	all := pl.offlinePeerIdentifiers()
+
+	var err error
+	err = multierr.Append(err, pl.updateOffline(peer.ListUpdates{
+		Removals: all,
+	}))
+	err = multierr.Append(err, pl.updateOnline(peer.ListUpdates{
+		Additions: all,
+	}))
+	return err
+}
+
+// Stop notifies the List that requests will stop coming
+func (pl *List) Stop() error {
+	return pl.once.Stop(pl.stop)
+}
+
+func (pl *List) stop() error {
+	pl.lock.Lock()
+	defer pl.lock.Unlock()
+
+	all := pl.onlinePeerIdentifiers()
+
+	var err error
+	err = multierr.Append(err, pl.updateOnline(peer.ListUpdates{
+		Removals: all,
+	}))
+	err = multierr.Append(err, pl.updateOffline(peer.ListUpdates{
+		Additions: all,
+	}))
+	return err
+}
+
+// IsRunning returns whether the peer list is running.
+func (pl *List) IsRunning() bool {
+	return pl.once.IsRunning()
+}
+
+// add retains a peer and sets up a thunk (a thin proxy for a peer) to receive
```

Comment:
I’ll grep out all the old thunks. Evidently not everyone’s heard of a thunk, so I went through and renamed them façade. It’s tricky because this is effectively a layer of middleware. On one side, they’re proxy peers, on the other they’re proxy subscribers.